efi/preinstall/check_fw_protections_intel.go

```diff
@@ -298,3 +298,36 @@ func checkPlatformFirmwareProtectionsIntelMEI(env internal_efi.HostEnvironment)
 	// Everything is ok
 	return nil
 }
+
+const bootGuardStatusMsr = 0x13a
+
+const (
+	bootGuardStatusTpmMask uint64 = (3 << 1)
+	bootGuardStatusTpmNone uint64 = (0 << 1)
+	bootGuardStatusTpm12   uint64 = (1 << 1)
+	bootGuardStatusTpm2    uint64 = (2 << 1)
+	bootGuardStatusTpmPtt  uint64 = (3 << 1)
+)
+
+func checkIsTpmDiscreteIntel(env internal_efi.HostEnvironmentAMD64) (discreteTPM bool, err error) {
+	bootGuardStatus, err := env.ReadMSRs(bootGuardStatusMsr)
+	if err != nil {
+		return false, fmt.Errorf("failed to read BootGuard status")
+	}
+	// NOTE: bootGuardStatus[0] is fine because BootGuard status MSR has the same value across all CPUs
+	switch bootGuardStatus[0] & bootGuardStatusTpmMask {
+	case bootGuardStatusTpmNone:
+		return false, fmt.Errorf("BootGuard status says the system has no TPM")
```

Comment:
In both of these cases, we have `ErrNoTPM2Device` so that the caller can detect what the error is - the error can be wrapped with more context using `fmt.Errorf` and the `%w` verb.